dagger/manifest.nim

```diff
@@ -0,0 +1,197 @@
+## Nim-Dagger
+## Copyright (c) 2021 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+{.push raises: [Defect].}
+
+import pkg/libp2p
+import pkg/libp2p/protobuf/minprotobuf
+import pkg/questionable
+import pkg/questionable/results
+import pkg/chronicles
+import pkg/chronos
+
+import ./blocktype
+
+const
+  ManifestCodec* = multiCodec("dag-pb")
+
+type
+  BlocksManifest* = object
+    blocks: seq[Cid]
+    htree: ?Cid
+    version*: CidVersion
+    hcodec*: MultiCodec
+    codec*: MultiCodec
+    cidEmptyDigest: MultiHash
```

Comment:
It feels a bit weird to me to have this as an instance variable. I think I would rather like to see a function that returns the empty digest, given a manifest, and drop the instance variable:

```nim
func emptyDigest(manifest: BlocksManifest): ?!MultiHash =
```

Reply:
The idea is to initialize `emptyDigest` only once. This was going to be originally done at compile time as a `const`, but `Cid.init()` uses `memmove` which breaks compile time initialization, so I had to cache it in the object itself. Maybe we can move it to a `template` instead or worst case scenario I can just add a TODO to mention the reasoning

Reply:
It seems that the empty digest depends on hcodec, so you'd have to find a way to specify that at compile time as well.

Reply:
Not really, this strings are self describing and you can initialize a `Cid` with them and it will initialize with the correct type of hash (hcodec) and content it points to, which in this case should be `raw`.

Reply:
> It seems that the empty digest depends on hcodec, so you'd have to find a way to specify that at compile time as well.

Hmm, you're right that we assume the hash to be `sha-256` and we'd need to be able to generate an empty digest for every possible hash type... thats a bit of a pain, maybe we're better off just restricting everything to `sha-256` for now?

Reply:
Restricting to SHA256 is fine by me.